src/datahike/query.cljc

```diff
@@ -90,14 +93,11 @@
 
 (defn distinct-tuples
   "Remove duplicates just like `distinct` but with the difference that it only works on values on which `vec` can be applied and two different objects are considered equal if and only if their results after `vec` has been applied are equal. This means that two different Java arrays are considered equal if and only if their elements are equal."
-  [tuples]
-  (first (reduce (fn [[dst seen] tuple]
-                   (let [key (vec tuple)]
-                     (if (seen key)
-                       [dst seen]
-                       [(conj dst tuple) (conj seen key)])))
-                 [[] #{}]
-                 tuples)))
+  ([tuples]
+   (into [] (distinct-tuples) tuples))
+  ([]
+   (let [seen (HashSet.)]
+     (filter #(.add ^HashSet seen (vec %))))))
```

Comment:
That's a good point! I wanted this function to be usable also as a transducer constructed from the expression `(distinct-tuples)` and this `(filter ...)` expression returns a filtering transducer that has a stateful predicate. In the two places where we use this transducer we use it eagerly so I don't think we will have any issues because of that.

I thought using a `HashSet` to implement this predicate was an elegant way of doing it. But I now realize it makes the code less portable to ClojureScript. So I can spend some time to see if we can have a pure Clojure solution without Java interop that does not sacrifice performance.

Reply:
That would be great!

Reply:
I came up with a pure Clojure solution that internally uses `clojure.core/distinct`.